rust/lance/src/io/exec/scalar_index_join.rs

```diff
@@ -0,0 +1,458 @@
+            }
+            datafusion::logical_expr::JoinType::Right => {
+                // Right join: Keep all source rows, _rowid can be null for unmatched rows
+                // This is effectively the same as Left join in our case since we're
+                // joining source (input) with target (index), and Right join means
+                // keep all rows from the right side (source/input)
+                (batch.columns().to_vec(), rowid_array)
+            }
+            _ => {
+                return Err(Error::NotSupported {
+                    source: format!("Join type {:?} not supported", self.join_type).into(),
+                    location: location!(),
+                });
+            }
+        };
+
+        // Build result batch
+        let mut result_columns = filtered_columns;
+        result_columns.push(filtered_rowid);
+
+        Ok(RecordBatch::try_new(
+            self.output_schema.clone(),
+            result_columns,
+        )?)
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for ScalarIndexJoinExec {
+    fn name(&self) -> &str {
+        "ScalarIndexJoinExec"
+    }
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn properties(&self) -> &PlanProperties {
+        &self.properties
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
+        vec![&self.input]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        if children.len() != 1 {
+            return Err(datafusion::error::DataFusionError::Internal(
+                "ScalarIndexJoinExec requires exactly one child".to_string(),
+            ));
+        }
+
+        Ok(Arc::new(Self::try_new(
+            children.into_iter().next().unwrap(),
+            self.dataset.clone(),
+            self.join_column.clone(),
+            self.index_name.clone(),
+            self.join_type,
+        )?))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> datafusion::error::Result<SendableRecordBatchStream> {
+        let input_stream = self.input.execute(partition, context)?;
+        let exec = self.clone();
+
+        let output_stream = input_stream
+            .map_err(Error::from)
+            .and_then(move |batch| {
+                let exec = exec.clone();
+                async move { exec.process_batch(batch).await }
+            })
+            .map_err(|e| datafusion::error::DataFusionError::External(Box::new(e)));
+
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            self.output_schema.clone(),
+            Box::pin(output_stream),
+        )))
+    }
+
+    fn statistics(&self) -> datafusion::error::Result<Statistics> {
+        // For now, return unknown statistics
+        // In the future, we could use index statistics to provide better estimates
```

Comment:
seems like we can provide some of these without index stats right? This feeds into analyze_plan I think.